// @pm2/io/build/main/features/eventLoopInspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const serviceManager_1 = require("../serviceManager");
const EventLoopInspector = require("event-loop-inspector");
const Debug = require("debug");
class EventLoopInspectorConfig {
}
class EventLoopInspectorFeature {
    constructor() {
        this.logger = Debug('axm:features:actions:eventloop');
    }
    init(config) {
        if (config === false)
            return;
        if (config === undefined)
            return;
        if (config === true) {
            config = { enabled: true };
        }
        if (config.enabled === false)
            return;
        this.actionService = serviceManager_1.ServiceManager.get('actions');
        if (this.actionService === undefined) {
            return this.logger('cannot expose actions as action service isnt available');
        }
        this.eventLoopInspector = EventLoopInspector(false);
        this.actionService.registerAction('km:event-loop-dump', (cb) => {
            return cb({
                success: true,
                dump: this.eventLoopInspector.dump()
            });
        });
    }
    destroy() {
        this.logger('destroy');
    }
}
exports.EventLoopInspectorFeature = EventLoopInspectorFeature;
