// @pm2/io/build/main/transports/IPCTransport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Debug = require("debug");
const eventemitter2_1 = require("eventemitter2");
const cluster = require("cluster");
const EventLoopInspector = require("event-loop-inspector");
class IPCTransport extends eventemitter2_1.EventEmitter2 {
    constructor() {
        super(...arguments);
        this.initiated = false;
        this.logger = Debug('axm:transport:ipc');
        this.eventLoopInspector = EventLoopInspector(false);
    }
    init(config) {
        this.logger('Init new transport service');
        if (this.initiated === true) {
            console.error(`Trying to re-init the transport, please avoid`);
            return this;
        }
        this.initiated = true;
        this.logger('Agent launched');
        this.onMessage = (data) => {
            this.logger(`Received reverse message from IPC`);
            this.emit('data', data);
        };
        process.on('message', this.onMessage);
        if (cluster.isWorker === false) {
            this.autoExitHook();
        }
        return this;
    }
    autoExitHook() {
        this.autoExitHandle = setInterval(() => {
            const dump = this.eventLoopInspector.dump();
            const requests = Object.keys(dump.requests);
            const handles = Object.keys(dump.handles);
            const sockets = dump.handles.Socket;
            const pipes = dump.handles.Pipe;
            const hasStdOut = sockets && sockets.find(sock => sock.fd === 1) !== undefined;
            const hasStdErr = sockets && sockets.find(sock => sock.fd === 2) !== undefined;
            const isSocketStds = sockets && sockets.length === 2 && hasStdErr && hasStdOut;
            const isOnlySocketStds = handles.length === 1 && isSocketStds;
            const isStdsWithIPC = isSocketStds && pipes && pipes.length === 1 && handles.length === 2;
            const shouldStillExit = isStdsWithIPC || isOnlySocketStds;
            if ((handles.length > 0 || requests.length > 0) && shouldStillExit === false) {
                return this.logger('no need to exit, there are handles/requests in uv', JSON.stringify(dump));
            }
            this.logger(`Nothing found in uv, removing the IPC listener`);
            process.removeListener('message', this.onMessage);
            let tmp = setTimeout(_ => {
                this.logger(`Still alive, listen back to IPC`);
                process.on('message', this.onMessage);
            }, 200);
            tmp.unref();
        }, 5000);
        this.autoExitHandle.unref();
    }
    setMetrics(metrics) {
        const serializedMetric = metrics.reduce((object, metric) => {
            if (typeof metric.name !== 'string')
                return object;
            object[metric.name] = {
                historic: metric.historic,
                unit: metric.unit,
                type: metric.id,
                value: metric.value
            };
            return object;
        }, {});
        this.send('axm:monitor', serializedMetric);
    }
    addAction(action) {
        this.logger(`Add action: ${action.name}:${action.type}`);
        this.send('axm:action', {
            action_name: action.name,
            action_type: action.type,
            arity: action.arity,
            opts: action.opts
        });
    }
    setOptions(options) {
        this.logger(`Set options: [${Object.keys(options).join(',')}]`);
        return this.send('axm:option:configuration', options);
    }
    send(channel, payload) {
        if (typeof process.send !== 'function')
            return -1;
        try {
            process.send({ type: channel, data: payload });
        }
        catch (err) {
            this.logger('Process disconnected from parent !');
            this.logger(err);
            return process.exit(1);
        }
    }
    destroy() {
        if (this.onMessage !== undefined) {
            process.removeListener('message', this.onMessage);
        }
        if (this.autoExitHandle !== undefined) {
            clearInterval(this.autoExitHandle);
        }
        this.logger('destroy');
    }
}
exports.IPCTransport = IPCTransport;
